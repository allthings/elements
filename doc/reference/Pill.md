<!-- 
This is an auto-generated markdown. 
You can change it in "/Users/daniel/Dev/allthings/elements/src/Pill/Pill.tsx" and run build:docs to update this file.
-->
# Pill
Pills ðŸ’Š are used to show status or to highlight a piece of information.
It could be used on cards to indicate that this card contains important information.

```example
<ThemeProvider>
   <View>
     <Pill label="Important message" color="primary" />
     <Pill label="A warning" color="warn" />
   </View>
</ThemeProvider>
```
## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|label **(required)**||Text that shows on the pill *
|color|string|Themed color of the pill *
