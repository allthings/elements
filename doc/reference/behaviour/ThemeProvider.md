<!-- 
This is an auto-generated markdown. 
You can change it in "src/behaviour/ThemeProvider.jsx" and run build:docs to update this file.
-->
# ThemeProvider
All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
The ThemeProvider allows you to define the default colors for most elements.

**Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|root|bool|
|theme|shape|<br>Default: {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.secondary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}
|children|node|
