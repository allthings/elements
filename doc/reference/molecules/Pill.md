<!-- 
This is an auto-generated markdown. 
You can change it in "src/molecules/Pill.jsx" and run build:docs to update this file.
-->
# Pill
Pills ðŸ’Š are used to show status or to highlight a piece of information.
It could be used on cards to indicate that this card contains important information.

```example
<ThemeProvider>
  <View>
    <Pill label="Important message" color="primary" />
    <Pill label="A warning" color="warn" />
  </View>
</ThemeProvider>
```
## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|label **(required)**|string|Text that shows on the pill *
|color|string|Themed color of the pill *<br>Default: 'primary'
