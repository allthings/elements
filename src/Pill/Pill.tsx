import React, { FunctionComponent } from 'react'
import { css } from 'glamor'

import Text from '../Text'
import Icon from '../Icon'
import { IViewProps } from '../View'
import Theme from '../Theme'
import Absolute from '../Absolute'
import Relative from '../Relative'

const styles = (backgroundColor: string, hasRemoveIcon: boolean) =>
  css({
    backgroundColor,
    borderRadius: hasRemoveIcon ? 10 : 10,
    cursor: 'default',
    height: 21,
    padding: '0 10px',
    paddingLeft: 10,
    paddingRight: hasRemoveIcon ? 25 : 10,
  })

/**
 * Pills ðŸ’Š are used to show status or to highlight a piece of information.
 * It could be used on cards to indicate that this card contains important information.
 *
 * ```example
 * <ThemeProvider>
 *   <View>
 *     <Pill label="Important message" color="primary" />
 *     <Pill label="A warning" color="warn" />
 *   </View>
 * </ThemeProvider>
 * ```
 */
const Pill: FunctionComponent<IPillProps> = ({
  color = 'primary',
  label,
  onRemoveClick,
  ...props
}) => {
  return (
    <Theme>
      {({ colorize }) => (
        <Relative
          alignH="center"
          alignV="center"
          direction="column"
          {...styles(colorize(color), !!onRemoveClick)}
          {...props}
        >
          <Text size="s" color="textOnBackground" strong block>
            {label}
          </Text>
          {onRemoveClick && (
            <Absolute
              right={6}
              top={-2}
              {...css({ padding: 2 })}
              onClick={onRemoveClick}
            >
              <Icon
                {...css({ margin: '0px 0px 5px 7px' })}
                name="remove-light-filled"
                color={colorize('white')}
                size={8}
              />
            </Absolute>
          )}
        </Relative>
      )}
    </Theme>
  )
}

interface IPillProps extends IViewProps {
  /** Text that shows on the pill **/
  label: string
  /** Themed color of the pill **/
  color?: string
  /** Themed color of the pill **/
  onRemoveClick?: () => void
}

export default Pill
