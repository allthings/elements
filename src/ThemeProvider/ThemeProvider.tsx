import React, { FunctionComponent, createContext } from 'react'
import { ColorPalette } from '@allthings/colors'

export const defaultTheme: ITheme = {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.secondary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}

export interface ITheme {
  readonly primary: string
  readonly text: string
  readonly secondaryText: string
  readonly titleColor: string
  readonly contrast: string
  readonly warn: string
  readonly disabled: string
  readonly background: string
  readonly textOnBackground: string
  readonly [key: string]: string
}

/**
 * All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
 * The ThemeProvider allows you to define the default colors for most elements.
 *
 * **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
 **/
const ThemeProvider: FunctionComponent<{ theme?: Partial<ITheme> }> = ({
  children,
  theme = defaultTheme,
}) => (
  <ThemeConsumer>
    {contextTheme => (
      <ThemeContext.Provider value={{ ...contextTheme, ...(theme as ITheme) }}>
        {children}
      </ThemeContext.Provider>
    )}
  </ThemeConsumer>
)

export default ThemeProvider
export const ThemeContext = createContext<ITheme>(defaultTheme)
export const ThemeConsumer = ThemeContext.Consumer
