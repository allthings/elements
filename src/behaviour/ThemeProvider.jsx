import React from 'react'
import PropTypes from 'prop-types'
import { Broadcast, Subscriber } from 'react-broadcast'
import { ColorPalette } from '@allthings/colors'

export const THEME_CHANNEL = 'theme'

export const defaultTheme = {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.primary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}

/**
* All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
* The ThemeProvider allows you to define the default colors for most elements.
*
* **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
**/
export default class ThemeProvider extends React.PureComponent {
  static propTypes = {
    root: PropTypes.bool,
    theme: PropTypes.shape({
      primary: PropTypes.string,
      text: PropTypes.string,
      secondaryText: PropTypes.string,
      titleColor: PropTypes.string,
      contrast: PropTypes.string,
      warn: PropTypes.string,
      disabled: PropTypes.string,
      background: PropTypes.string,
      textOnBackground: PropTypes.string,
    }),
    children: PropTypes.node,
  }

  static defaultProps = {
    theme: defaultTheme,
  }

  static contextTypes = Subscriber.contextTypes

  render() {
    let root = true
    if (this.context.broadcasts && this.context.broadcasts[THEME_CHANNEL]) {
      root = false
    }
    const { theme, children } = this.props
    return root ? (
      <Broadcast
        channel={THEME_CHANNEL}
        value={{
          ...ThemeProvider.defaultProps.theme,
          ...theme,
        }}
      >
        {children}
      </Broadcast>
    ) : (
      <Subscriber channel={THEME_CHANNEL}>
        {previousTheme => (
          <Broadcast
            channel={THEME_CHANNEL}
            value={{
              ...ThemeProvider.defaultProps.theme,
              ...previousTheme,
              ...theme,
            }}
          >
            {children}
          </Broadcast>
        )}
      </Subscriber>
    )
  }
}

export const ThemeShape = ThemeProvider.propTypes.theme

export const withTheme = (mapThemeToProps, displayName) => WrappedComponent =>
  // eslint-disable-next-line
  class extends React.PureComponent {
    static displayName = displayName || WrappedComponent.displayName
    static component = WrappedComponent

    renderComponent = theme => {
      const props = !mapThemeToProps
        ? { theme }
        : mapThemeToProps(theme, this.props)

      return <WrappedComponent {...this.props} {...props} />
    }

    render() {
      return (
        <Subscriber channel={THEME_CHANNEL}>{this.renderComponent}</Subscriber>
      )
    }
  }
