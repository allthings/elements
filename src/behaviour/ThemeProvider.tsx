import React, { FunctionComponent, createContext } from 'react'
import { ColorPalette } from '@allthings/colors'

export const defaultTheme = {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.secondary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}

export interface ITheme {
  primary: string
  text: string
  secondaryText: string
  titleColor: string
  contrast: string
  warn: string
  disabled: string
  background: string
  textOnBackground: string
  [key: string]: string
}

const ThemeContext = createContext<ITheme>(defaultTheme)
export const ThemeConsumer = ThemeContext.Consumer

/**
 * All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
 * The ThemeProvider allows you to define the default colors for most elements.
 *
 * **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
 **/
const ThemeProvider: FunctionComponent<{ theme?: Partial<ITheme> }> = ({
  children,
  theme = {},
}) => (
  <ThemeConsumer>
    {contextTheme => (
      <ThemeContext.Provider value={{ ...contextTheme, ...(theme as ITheme) }}>
        {children}
      </ThemeContext.Provider>
    )}
  </ThemeConsumer>
)

export default ThemeProvider

type mapThemeToPropsType = (theme: ITheme, props: any) => any

function createThemeConsumer<T>(
  mapThemeToProps: mapThemeToPropsType,
  WrappedComponent: React.ComponentType<T>,
  props: Omit<T, 'theme'>
) {
  return (theme: ITheme) => {
    const extraProps = !mapThemeToProps
      ? { theme }
      : mapThemeToProps(theme, props)

    return <WrappedComponent {...props} {...extraProps} />
  }
}

export function withTheme(mapThemeToProps: mapThemeToPropsType) {
  return function<T extends { theme: ITheme }>(
    WrappedComponent: React.ComponentType<T>
  ): React.ComponentType<Omit<T, 'theme'>> {
    return function(props: Omit<T, 'theme'>) {
      return (
        <ThemeConsumer>
          {createThemeConsumer<T>(mapThemeToProps, WrappedComponent, props)}
        </ThemeConsumer>
      )
    }
  }
}
