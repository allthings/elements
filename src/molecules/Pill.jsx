import React from 'react'
import PropTypes from 'prop-types'
import { css } from 'glamor'

import Text from '../atoms/Text'
import View from '../atoms/View'
import Theme from '../behaviour/Theme'

const styles = backgroundColor =>
  css({
    backgroundColor,
    display: 'inline-block',
    borderRadius: 10,
    padding: '2px 15px',
    cursor: 'default',
  })

const Pill = ({ color = 'primary', label, ...props }) => {
  return (
    <Theme>
      {({ colorize }) => (
        <View {...styles(colorize(color))} {...props}>
          <Text size="s" color="textOnBackground" strong>
            {label}
          </Text>
        </View>
      )}
    </Theme>
  )
}

Pill.propTypes = {
  label: PropTypes.string.isRequired,
  color: PropTypes.string,
}

export default Pill
