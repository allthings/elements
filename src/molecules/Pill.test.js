import React from 'react'
import renderer from 'react-test-renderer'
import Pill from './Pill'
import ThemeProvider from '../behaviour/ThemeProvider'

test('Pill renders without error', () => {
  const tree = renderer
    .create(
      <ThemeProvider>
        <Pill label="Important message" />
      </ThemeProvider>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
