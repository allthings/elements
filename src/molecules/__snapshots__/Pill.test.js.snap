// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pill renders without error 1`] = `
.css-9j98ks,
[data-css-9j98ks] {
  background-color: #3598db;
  display: inline-block;
  border-radius: 10px;
  padding: 2px 15px;
  cursor: default;
}

.css-4p3ux5,
[data-css-4p3ux5] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.css-7xie1s,
[data-css-7xie1s] {
  display: block;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-weight: 600;
  font-size: 12px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #ffffff;
}

<div
  data-css-4p3ux5=""
  data-css-9j98ks=""
>
  <div
    data-css-4p3ux5=""
    data-css-7xie1s=""
  >
    Important message
  </div>
</div>
`;
